### PR TITLE
fix `subpath` filtering for `AzureDevOps` `readTree`

### PR DESCRIPTION
**Pull request by benjdlambert**
There's a root directory that we need to strip here when using `readTree` for `AzureDevOps`. This should hopefully fix the templating bug that we saw in #5708.

### CODE REVIEW

**Comment by benjdlambert** [comment]
@awanlin can you check out this branch and test with some repo's. This should fix the scaffolding, but there's a possibility that it could break techdocs. Want to confirm if it breaks for you, and what the error message is so we can fix that up before shipping this.

**Comment by awanlin** [comment]
@benjdlambert I managed to get some testing in now, so far just the TechDocs with the Basic out of box setup. I'm getting these errors:

In the header area: Building a newer version of this documentation failed. Error: "Failed to generate docs from C:\\Users\\awanlin\\AppData\\Local\\Temp\\backstage-VQZMbt into C:\\Users\\awanlin\\AppData\\Local\\Temp\\techdocs-tmp-Fphch2 with error Docker container returned a non-zero exit code (1)"

In the Build Details side panel logs: "ERROR    -  Config value: 'docs_dir'. Error: The path /input/Documentation isn't an existing directory.
Aborted with 1 Configuration Errors!
error: Failed to build the docs page: Failed to generate docs from C:\Users\awanlin\AppData\Local\Temp\backstage-VQZMbt into C:\Users\awanlin\AppData\Local\Temp\techdocs-tmp-Fphch2 with error Docker container returned a non-zero exit code (1) {"timestamp":"2021-09-07T12:34:01.211Z"}"

Looking at the listed directory I for sure can tell that the repo isn't in the proper state and the stripping is for sure the culprit. This is the same file output I had seen previously with #4359

**Comment by awanlin** [comment]
The scaffolder side works fine though, I get the expected repo structure there. Currently in master you would get `root-folder-01` as a folder and then everything under it. With this branch we do not which is what would be ideal, here's a screenshot of that:

image

**Comment by benjdlambert** [comment]
@awanlin perfect, ok thanks for this I can try and have a play around to see if we can replicate this and try and fix it, looks like it should be pretty easy to fix though, now it's got the correct behaviour, it looks like the TechDocs implementation is wrong for Azure.

File: packages/backend-common/src/reading/tree/ReadTreeResponseFactory.ts [hunk 1]
```diff
@@ -55,6 +55,7 @@ export class DefaultReadTreeResponseFactory implements ReadTreeResponseFactory {
       this.workDir,
       options.etag,
       options.filter,
+      options.stripFirstDirectory,
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply packages/backend-common/src/reading/tree/ReadTreeResponseFactory.ts:58]: Yep good call. I was thinking of splitting the types but maybe that's more hassle than it's worth.

**Comment by benjdlambert** [comment]
@awanlin we worked it out, we were not filtering the contents of the `zip` directory which is returned from `AzureDevOps`.

**Review by freben — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]
:shipit:, but as mentioned let's get rid of the `stripFirstDirectory` option too

**Comment by awanlin** [comment]
Finally got around to testing this with Azure DevOps Server, the on-prem version, and I'm getting this:

image

We should see only what is in the green box and the items in the red boxes should not be there. I've tripled checked to make sure I've got the latest code and I do.

When I debug this `filePath` only gets hit once and it's an empty string: https://github.com/backstage/backstage/blob/36e26fb9591cd62a87876bec39329d3f24cbd952/packages/backend-common/src/reading/AzureUrlReader.ts#L130 Is that what we should expect?

**Comment by Rugvip** [comment]
@awanlin very likely because of this https://github.com/IonicaBizau/git-url-parse/blob/246c9119fb42c2ea1c280028fe77c53eb34c190c/lib/index.js#L106

@benjdlambert I think we should just parse it like a normal url and grab the path param instead

**Comment by Rugvip** [comment]
Realizing that the handling of azure URLs make the same assumption throughout the entire project, so not gonna try to sneak a fix in for this before the release today : /

Just to verify, you're hosting azure on a completely different domain than `dev.azure.com` right @awanlin?

**Comment by Rugvip** [comment]
Well jeez, we don't want to ship this either. For the azure readTree we use the `getAzureDownloadUrl` function to create the download URL, which actually takes care or selecting the correct subfolder in the repo. With this change I think we're actually about to break the behaviour for all azure users 😅

The `getAzureDownloadUrl` doesn't handle on-prem deployments properly though, which is probably why we're been seeing issues.

I'll verify if ^ is correct and in that case roll back this so we can get the release out, aka taking the red pill and joining Ben in the rabbit hole. Then perhaps also set something up so we can test this e2e for both azure cloud and on-prem ._.

**Comment by awanlin** [comment]
@Rugvip yes, we host it ourselves and the URLs will not use `dev.azure.com`. They can be pretty much anything when you host it yourself. I did recently submit a PR that fixes the `getAzureDownloadUrl` for our use case as it was breaking both the TechDocs and Scaffolder features very early in each of those.

https://github.com/backstage/backstage/blob/c6b809c6255b2c0d70f8f8905685108a4008e89b/packages/integration/src/azure/core.ts#L107

**Pull request merged**